### PR TITLE
Classifier for ISIC 2020 Kaggle Challenge based on a Siamese Network 

### PR DESCRIPTION
**Pull request by LimaoC**
## Classifier for ISIC 2020 Kaggle Challenge based on a Siamese Network

This PR adds an implementation of a Siamese network based classifier for the ISIC 2020 Melanoma Classification Challenge. For more details, see the README.

## Files
- `README.md`: Documentation of problem description and model architecture, dependencies, results, training details, and references.
- `assets/`: Figures used in README
- `environment.yml`: Conda environment file for reproducibility of environment and dependencies
- `src/dataset.py`: Functionality for loading and preprocessing data
- `src/modules.py`: Components for Siamese network and custom MajorityClassifier
- `src/predict.py`: Example usage of trained model
- `src/train.py`: Code for training, validating, testing, and saving model
- `src/util.py`: Various plotting utilities

Submission by Limao Chang (46978554).

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **20**

Solves problem: The solution is well-suited to the problem (**5**).

multiple plots and performance parameters are provided
The Siamese network approach with a ResNet-50 backbone and contrastive loss
PCA and t-SNE plots suggest limited separability in the learned embedding space, with some clustering evident, particularly in the malignant predictions.
Class imbalance is addressed through undersampling of benign cases.
AUROC = 88%,  Sensitivity= 67%, and specificity 87% with Margin = 0.5 and learning rate = 0.00001.
Implementation functions as intended (**3**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

Note:

- Using an algorithmic approach to undersample the majority class is indeed recommended over random sampling, especially in the context of medical imaging to maintain informative samples, improve generalization, and avoide loss of critical information.

- The model demonstrates high specificity and overall accuracy but struggles with sensitivity, likely due to the challenge of learning effective features for malignant cases.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splits. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>0</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Great work!
